File: leetcode/medium/1339. Maximum Product of Splitted Binary Tree/main.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    _MOD = 10 ** 9 + 7

    def maxProduct(self, root: Optional[TreeNode]) -> int:
        result = 0
        totalSum = self.treeSum(root)

        def traverse(node: Optional[TreeNode]) -> int:
            nonlocal result

            if node is None:
                return 0

            leftSum = traverse(node.left)
            rightSum = traverse(node.right)

            subtreeSum = node.val + leftSum + rightSum
            result = max(result, subtreeSum * (totalSum - subtreeSum))

            return subtreeSum

        traverse(root)

        return result % Solution._MOD

    def treeSum(self, node: Optional[TreeNode]) -> int:
        if node is None:
            return 0
        return node.val + self.treeSum(node.left) + self.treeSum(node.right)

File: leetcode/medium/1339. Maximum Product of Splitted Binary Tree/main.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}


function treeSum(node: TreeNode | null): number {
  if (node === null) {
    return 0
  }

  return node.val + treeSum(node.left) + treeSum(node.right)
}


function maxProduct(root: TreeNode | null): number {
  const MOD = 1000000000 + 7

  let [totalTreeSum, result] = [treeSum(root), Number.MIN_VALUE]

  const traverse = (node: TreeNode | null): number => {
    if (node === null) {
      return 0
    }

    let [leftSum, rightSum] = [traverse(node.left), traverse(node.right)]

    result = Math.max(result, leftSum * (totalTreeSum - leftSum))
    result = Math.max(result, rightSum * (totalTreeSum - rightSum))

    return node.val + leftSum + rightSum
  }

  traverse(root)

  return result % MOD
};
